### ISSUE TITLE
Unexpected interaction: haskell-process-load-file/haskell-interactive-mode-run-expr 

### ISSUE THREAD

**Issue by thechriskind** [haskell/haskell-mode#941]
haskell-mode version: 20151015

The following function when run in a buffer editing a Haskell file, say Main.hs, inserts a newline in Main.hs.  If I remove the call to haskell-process-load-file, the problem goes away, but I lose the load-file functionality.  The other load file functions behave similarly.  

In my haskell-mode-hook, the only setup is calling  (interactive-haskell-mode).  I have tested this with emacs -q to verify that it's not an interaction with other state in my environment.

    (defun haskell-run-main ()
      (interactive)
      (haskell-process-load-file)
      (let* ((session (haskell-interactive-session))
             (buffer (haskell-session-interactive-buffer session))
          (with-current-buffer buffer
             (haskell-interactive-mode-run-expr "main"))))

**Comment by gracjan** [comment]
Good catch. Can you help to debug this further?

**Comment by thechriskind** [comment]
Thanks for the quick response.  I am happy to look into it, though it may take some time.  I know Clojure well, but my elisp and common-lisp are fairly limited.  However I am happy to learn and to help.

**Comment by thechriskind** [comment]
Apparently  haskell-interactive-mode-run-expr assumes that in commands queued using haskell-process-queue-command, the :go function will be run in the interactive buffer.  So it calls (insert "\n") in the :go function.

The interactive mode architecture document on the wiki does not state any such guarantee, and unlike :live, the :go function is not passed a buffer. Hence I believe that assumption is not valid.

**Comment by gracjan** [comment]
Should the :go function run in the context of that buffer or not?

**Comment by thechriskind** [comment]
Yes, the :go function in haskell-interactive-mode-run-expr should run in the context of the haskell session's interactive buffer.  The lines that make that assumption are

    (goto-char (point-max))
    (insert "\n")

**Comment by gracjan** [comment]
@thechriskind, we have two ways to fix this:
- either add the guarantee that `:go` functions are called in interactive buffer context
- or add `with-current-buffer` to the place where it is needed

Can you check which of the two approaches is more appropriate for haskell-interactive-mode and report here back? I'd like to see and hear what other people involved in this project think about this architectural decision.
